src/Compilers/CSharp/Test/Semantic/Semantics/StructConstructorTests.cs

```diff
@@ -0,0 +1,1669 @@
+@"{
+  // Code size       15 (0xf)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  ldnull
+  IL_0002:  call       ""S2..ctor(object)""
+  IL_0007:  ldarg.0
+  IL_0008:  ldc.i4.1
+  IL_0009:  stfld      ""bool S2.Initialized""
+  IL_000e:  ret
+}");
+            verifier.VerifyIL("S2..ctor(object)",
+@"{
+  // Code size        8 (0x8)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  ldc.i4.0
+  IL_0002:  stfld      ""bool S2.Initialized""
+  IL_0007:  ret
+}");
+        }
+
+        [Fact]
+        public void ThisInitializer_03()
+        {
+            var source =
+@"using System;
+struct S0
+{
+    internal bool Initialized = true;
+    internal S0(object obj) : this() { }
+}
+struct S1
+{
+    internal bool Initialized = true;
+    public S1() { }
+    internal S1(object obj) : this() { }
+}
+struct S2
+{
+    internal bool Initialized = true;
+    public S2() : this(null) { }
+    S2(object obj) { }
+}
+class Program
+{
+    static void Main()
+    {
+        Console.WriteLine(new S0().Initialized);
+        Console.WriteLine(new S0(null).Initialized);
+        Console.WriteLine(new S1().Initialized);
+        Console.WriteLine(new S1(null).Initialized);
+        Console.WriteLine(new S2().Initialized);
+    }
+}";
+
+            var verifier = CompileAndVerify(source, parseOptions: TestOptions.RegularPreview, expectedOutput:
+@"False
+False
+True
+True
+True");
+            verifier.VerifyMissing("S0..ctor()");
+            // PROTOTYPE: S0.Initialized should be set after initobj, not before, and
+            // expectedOutput should be False, True, ...
+            verifier.VerifyIL("S0..ctor(object)",
+@"{
+  // Code size       15 (0xf)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  ldc.i4.1
+  IL_0002:  stfld      ""bool S0.Initialized""
+  IL_0007:  ldarg.0
+  IL_0008:  initobj    ""S0""
+  IL_000e:  ret
+}");
+            verifier.VerifyIL("S1..ctor()",
+@"{
+  // Code size        8 (0x8)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  ldc.i4.1
+  IL_0002:  stfld      ""bool S1.Initialized""
+  IL_0007:  ret
+}");
+            verifier.VerifyIL("S1..ctor(object)",
+@"{
+  // Code size        7 (0x7)
+  .maxstack  1
+  IL_0000:  ldarg.0
+  IL_0001:  call       ""S1..ctor()""
+  IL_0006:  ret
+}");
+            verifier.VerifyIL("S2..ctor()",
+@"{
+  // Code size        8 (0x8)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  ldnull
+  IL_0002:  call       ""S2..ctor(object)""
+  IL_0007:  ret
+}");
+            verifier.VerifyIL("S2..ctor(object)",
+@"{
+  // Code size        8 (0x8)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  ldc.i4.1
+  IL_0002:  stfld      ""bool S2.Initialized""
+  IL_0007:  ret
+}");
+        }
+
+        [Fact]
+        public void ThisInitializer_04()
+        {
+            var source =
+@"using System;
+struct S0
+{
+    internal bool Initialized = false;
+    internal S0(object obj) : this() { Initialized = true; }
+}
+struct S1
+{
+    internal bool Initialized = false;
```

Comment:
>Initialized [](http://example.com/codeflow?start=18&length=11)

Consider using a different type here. I can't tell whether `new S1().Initialized` ran the field initializer or not.

Reply:
This is a general comment for all the `ThisInitializer` tests, I can't reason about the behavior of the existing code.

Reply:
@cston this appears unaddressed.

Reply:
I've added printing to the previous test to indicate the order of evaluation. This test is similar to the previous test but in this case is verifying the decision whether to execute the initializer is independent of the initialized value, so it was necessary for the initializers to use the default values. (I've added a comment to this test to indicate the intent.)